ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -117,76 +127,35 @@ delayedProgressReporting before after (Just lspEnv) optProgressStyle = do
     progressState <- newVar NotStarted
     let progressUpdate event = updateStateVar $ Event event
         progressStop   =  updateStateVar StopProgress
-        updateStateVar = modifyVar_ progressState . updateState (lspShakeProgress inProgressState)
-
+        updateStateVar tran = do
+            start <- lspShakeProgressNew inProgressState
+            modifyVar_ progressState $ updateState start tran
         inProgress = updateStateForFile inProgressState
     return ProgressReporting{..}
     where
-        lspShakeProgress InProgressState{..} = do
+        lspShakeProgressNew InProgressState{..} = do
             -- first sleep a bit, so we only show progress messages if it's going to take
             -- a "noticable amount of time" (we often expect a thread kill to arrive before the sleep finishes)
             liftIO $ sleep before
-            u <- ProgressToken . InR . T.pack . show . hashUnique <$> liftIO newUnique
-
-            b <- liftIO newBarrier
-            void $ LSP.runLspT lspEnv $ LSP.sendRequest SMethod_WindowWorkDoneProgressCreate
-                LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
-            liftIO $ async $ do
-                ready <- waitBarrier b
-                LSP.runLspT lspEnv $ for_ ready $ const $ bracket_ (start u) (stop u) (loop u 0)
+            cancelProgress <- newBarrier
+            LSP.runLspT lspEnv $ withProgress "Processing" Nothing Cancellable $ \update ->
+                race (liftIO $ waitBarrier cancelProgress) (loop update 0)
+            return cancelProgress
```

Comment:
This can't work: `withProgress` runs until the inner operation terminates! So we won't ever get here!

Reply:
(Sorry, I don't have a suggestion for how to do it right, I need to read it carefully again...)

Reply:
Right, I forget to async it. I'll add it and see how would this go